### PR TITLE
PyTorch Datasets

### PR DESCRIPTION
**Pull request by jvmancuso**
Anticipating a time when we have processed data from the Kaggle data, I've begun mocking up what I think we'll need in order to interface with PyTorch.  To understand how this interrelates, it's important to view the tables and their relations as a data table tree with the following schema:
```
application_{train/test}.csv
    bureau.csv
        bureau_balance.csv
    previous_applications.csv
        POS_cash_balance.csv
        installments_payments.csv
        credit_card_balance.csv
```

The assumption I've been making is that each parent dataset (e.g. application) will have a mapping for any of it's children datasets (e.g. bureau) that maps each of its own indices to a collection of indices in the child.  Because we need to use these mappings from the top level RiskDataset, this means that the leaf node Datasets (e.g. bureau balance) will need to include a nested mapping that maps from root index -> child indices -> leaf indices (for example, Application index -> Bureau indices -> BureauBalance indices).  We then build a custom PyTorch Sampler that uses these mappings to ensure that the correct instances are retrieved from the child dataset.  I've implemented the first one for Bureau and BureauBalance, and the others should be done analogously (in sequence).

This is the best approach I can think of, as it allows us to take advantage of all of PyTorch's other utilities (e.g. DataLoader).  If there's a simpler way of doing this, I'd love to hear about it :D

### CODE REVIEW

File: defaultrisk/core/dataset.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+import torch
+from torch.utils.data import Dataset
+
+# After collating, Dimensions of each instance in RiskDataset
```
Comment (bmorphism) [inline defaultrisk/core/dataset.py]: Very helpful!
